test/test_utils_log.cpp

```diff
@@ -0,0 +1,352 @@
+
+                        helper_log_init(envVar.c_str());
+                        helper_checkConfig(&b, &loggerConfig);
+                    }
+                }
+            }
+        }
+    }
+}
+
+// Unfortunately this must be a macro to pass variadic argument
+#define helper_test_log(...)                                                   \
+    do {                                                                       \
+        fput_count = 0;                                                        \
+        fflush_count = 0;                                                      \
+        utils_log(__VA_ARGS__);                                                \
+        EXPECT_EQ(fput_count, expect_fput_count);                              \
+        EXPECT_EQ(fflush_count, expect_fflush_count);                          \
+    } while (0)
+
+static std::string helper_log_str(int l) {
+    switch (l) {
+    case LOG_DEBUG:
+        return "DEBUG";
+    case LOG_ERROR:
+        return "ERROR";
+    case LOG_INFO:
+        return "INFO ";
+    case LOG_WARNING:
+        return "WARN ";
+    default:
+        ASSERT(0);
+    }
+}
+
+TEST_F(test, log_levels) {
+    expected_stream = stderr;
+    for (int i = LOG_DEBUG; i <= LOG_ERROR; i++) {
+        for (int j = LOG_DEBUG; j <= LOG_ERROR; j++) {
+            loggerConfig = {0, 0, (LogLevel)i, LOG_DEBUG, stderr};
+            if (i > j) {
+                expect_fput_count = 0;
+                expect_fflush_count = 0;
+                expected_message = "";
+            } else {
+                expect_fput_count = 1;
+                expect_fflush_count = 1;
+            }
+            expected_message = "[" + helper_log_str(j) + " UMF] example log\n";
+            helper_test_log((LogLevel)j, "example log");
+        }
+    }
+}
+
+TEST_F(test, log_outputs) {
+    std::vector<FILE *> outs = {stdout, stderr, MOCK_FILE_PTR};
+    expect_fput_count = 1;
+    expect_fflush_count = 1;
+    expected_message = "[DEBUG UMF] example log\n";
+    for (auto o : outs) {
+        loggerConfig = {0, 0, LOG_DEBUG, LOG_DEBUG, o};
+        expected_stream = o;
+        helper_test_log(LOG_DEBUG, "example log");
+    }
+}
+
+TEST_F(test, flush_levels) {
+    expected_stream = stderr;
+    expect_fput_count = 1;
+    for (int i = LOG_DEBUG; i <= LOG_ERROR; i++) {
+        for (int j = LOG_DEBUG; j <= LOG_ERROR; j++) {
+            loggerConfig = {0, 0, LOG_DEBUG, (LogLevel)i, stderr};
+            if (i > j) {
+                expect_fflush_count = 0;
+            } else {
+                expect_fflush_count = 1;
+            }
+            expected_message = "[" + helper_log_str(j) + " UMF] example log\n";
+            helper_test_log((LogLevel)j, "example log");
+        }
+    }
+}
+
+TEST_F(test, long_log) {
+    expect_fput_count = 1;
+    expect_fflush_count = 1;
+    loggerConfig = {0, 0, LOG_DEBUG, LOG_DEBUG, stderr};
+    expected_message = "[DEBUG UMF] " LONG_STRING_2047 "\n";
```

Comment:
nit: instead of having all those LONG_STRING defines you can just do:
```
"[DEBUG UMF] " + std::string(2047, 'x') +  "\n"
``` 
or just make a function that returns std::string of specified length if you need to have specific content

Reply:
Done